File: first_bad_version/first_bad_version.go
package main

// Mocking the API isBadVersion
func isBadVersion(version int) bool {
    return version >= 4 // Simulating that version 4 and above are bad
}

func firstBadVersion(n int) int {
    left, right := 1, n

    for left < right {
        mid := left + (right-left)/2

        if isBadVersion(mid) {
            right = mid // The bad version is on the left side or is mid itself
        } else {
            left = mid + 1 // The bad version is on the right side
        }
    }

    return left
}

package main

import (
	"testing"
)

func TestFirstBadVersion(t *testing.T) {
	// Define a helper function to mock different cases
	tests := []struct {
		n        int
		expected int
	}{
		{5, 4},
		{10, 4},
		{7, 4},
		{3, 3},  // Simulating first bad version is 3
		{1000, 4}, // Large input test case
	}

	for _, test := range tests {
		result := firstBadVersion(test.n)
		if result != test.expected {
			t.Errorf("firstBadVersion(%d) = %d; expected %d", test.n, result, test.expected)
		}
	}
}

File: first_bad_version/first_bad_version.exs
defmodule Solution do
  # Mocking the API is_bad_version/1
  def is_bad_version(version) do
    version >= 4 # Simulating that version 4 and above are bad
  end

  def first_bad_version(n), do: binary_search(1, n)

  defp binary_search(left, right) when left < right do
    mid = div(left + right, 2)

    if is_bad_version(mid) do
      binary_search(left, mid) # Move left if mid is bad
    else
      binary_search(mid + 1, right) # Move right otherwise
    end
  end

  defp binary_search(left, _right), do: left # Return first bad version
end
